### ISSUE TITLE
Memory Error

### ISSUE THREAD

**Issue by Huii** [angus924/rocket#3]
First of all: Thank you very much for providing this new method! If it works, it works very well and leads to good results (i.e. leads to results comparable to other methods within minutes instead of weeks).

Unfortunately, it doesn't work all the time for me. Especially if I use 10,000 kernels as recommended in your paper. This is what I end up with on a Windows 10 machine equipped with an NVIDIA 1080 TI GPU and 32 GB RAM:

```
Algorithms.trainAndEvaluateROCKET(...)
  File "PathToMyProject\MyClass.py", line 180, in trainAndEvaluateROCKET
    classifier.fit(X_training_transform,trainY)
  File "PathToMyAnacondaFolder\lib\site-packages\sklearn\linear_model\ridge.py", line 1815, in fit
    _BaseRidgeCV.fit(self, X, Y, sample_weight=sample_weight)
  File "PathToMyAnacondaFolder\lib\site-packages\sklearn\linear_model\ridge.py", line 1528, in fit
    estimator.fit(X, y, sample_weight=sample_weight)
  File "PathToMyAnacondaFolder\lib\site-packages\sklearn\linear_model\ridge.py", line 1436, in fit
    X_mean, *decomposition = decompose(X, y, sqrt_sw)
  File "PathToMyAnacondaFolder\lib\site-packages\sklearn\linear_model\ridge.py", line 1348, in _svd_decompose_design_matrix
    U, s, _ = linalg.svd(X, full_matrices=0)
  File "PathToMyAnacondaFolder\lib\site-packages\scipy\linalg\decomp_svd.py", line 129, in svd
    full_matrices=full_matrices, overwrite_a=overwrite_a)
MemoryError
```

My testing data-set consists of 3,278 time series (with a length of 1041 amplitude values each) and the training data-set consists of 43,264 time series (again with a length of 1041 amplitude values each). Any help is highly appreciated.

// edit:I forgot to mention that it works perfectly fine for 100 or 1,000 kernels.

**Comment by angus924** [comment]
Hi Huii,

Thank you very much for your query.  You have touched on a very important issue.

**Solution**

The solution is to use ROCKET with logistic / softmax regression (instead of with the ridge regression classifier).  See [`reproduce_experiments_scalability.py`](https://github.com/angus924/rocket/blob/master/code/reproduce_experiments_scalability.py) for an example of how to integrate ROCKET with logistic / softmax regression using PyTorch.

Also, as you have a GPU, while we do not have an 'official' GPU implementation (at least, not yet), you might want to consider using or adapting the GPU implementation of ROCKET developed by @oguiza (see this [notebook](https://github.com/timeseriesAI/timeseriesAI/blob/master/05_ROCKET_a_new_SOTA_classifier.ipynb)).  (However, while I don't doubt that it works, I have not yet used or tested this implementation myself.)

**Explanation**

The issue you are experiencing is (almost certainly) due to a combination of:

* the size of your training set; and
* the relatively high memory complexity of the ridge regression classifier (RidgeClassifierCV).

The total size of the transformed features in your case, for 43K training examples, would be about 7 GB.  So, the issue is not the size of the transformed features per se, but rather the limits of the ridge regression classifier.

We suggest that for larger datasets (broadly speaking, where the number of training examples >> the number of features), ROCKET should be used with logistic / softmax regression.  This is our approach for our own scalability experiments.  The basic idea is to integrate ROCKET with minibatch stochastic gradient descent (or, e.g., Adam): perform the transform on each tranche / minibatch, and use the transformed tranches / minibatches for training.  You can cache the results of the transform (in your case, you should be able to easily cache the whole transform) so that you don't needlessly repeat the transform for multiple epochs of training.  However, with 43K training examples, it is entirely possible that a single pass / epoch will be sufficient for training / validation loss to converge anyway (i.e., caching may not be necessary), but this will depend on your dataset.  You may also find that with a large number of training examples, logistic / softmax regression is faster than the ridge regression classifier anyway.  The basic advantage of using this set up is that the memory cost becomes fixed (proportional to tranche / minibatch size), so memory consumption should no longer be an issue.

In case it is of any use, also see sections 3.3, 3.4.2, and 4.2.1 of our [paper](https://arxiv.org/pdf/1910.13051.pdf).

(Note that time series length is irrelevant to memory consumption, as ROCKET produces the same number of features, for a given number of kernels, regardless of time series length: by default, 20,000 features.)

I'm working on updated documentation which hopefully will make all this clear.

Please don't hesitate to let me know if this doesn't help you, or you need more information.  I'd be happy to try to help you to adapt the code in [`reproduce_experiments_scalability.py`](https://github.com/angus924/rocket/blob/master/code/reproduce_experiments_scalability.py) to work for your dataset.

I hope this helps!

And also, thanks for using ROCKET!

Best,

Angus.

**Comment by Huii** [comment]
Thank you very much for your very comprehensive answer. It's very helpful. :-)

I'll have a look at the GPU implementation of  @oguiza and your code as well. But before, just a quick follow-up question: Is it possible to use the logistic regression by simply substituting the following line:
`classifier = RidgeClassifierCV(alphas = np.logspace(-3,3,10), normalize = True)`
with
`classifier = LogisticRegression(...)`
and if yes, what parameters would you recommend?

I'm just asking this because by using the LogisticRegression method of scikit-learn, I wouldn't have to install PyTorch.

**Comment by angus924** [comment]
Broadly speaking, you should be able to use any classifier with ROCKET.  So, in principle, scikit-learn LogisticRegression (or SGDClassifier) would work.

I think the two issues in practice in your case are: (1) is the classifier fast enough; and (2) will the classifier work with the available memory.  I haven't used LogisticRegression (or SGDClassifier) recently, so I can only provide limited advice.  It might be worth trying (perhaps on a subset of your data) to see how well, and how fast, it will work.  My main suggestions in terms of hyperparameters would be to use the 'sag' or 'saga' solver for LogisticRegression, and to start with a small value for `max_iter`. Also, normalise the transformed features by subtracting the per-feature mean, and dividing by the per-feature standard deviation, before fitting the classifier (similar per-feature normalisation is handled internally by RidgeClassifierCV, but not by LogisticRegression).

However, I would recommend having a go at adapting our PyTorch code for your dataset.  It has been written to deal with more or less the same issue you are facing, and I am confident that it works (I've used it for repeated experiments on > 1 million time series).  Although PyTorch syntax / use is quite different to scikit-learn, you are unlikely have to change anything substantive in the code to make it work for you. You may need to adjust how the data is loaded, depending on the format and structure of your dataset, and to adjust the `cache_size` argument (as I noted above, you can probably set this to cover your entire training set).  I suspect it will be significantly faster than the scikit-learn classifiers.

Let me know how you go.

Best,

Angus.

**Comment by Huii** [comment]
Thank you once again for your very helpful reply. I decided to go with PyTorch and ended up writing the following code (serving as a wrapper for your methods):

```
    numberOfKernels = 10 #Setting a small value just for debugging
    numberOfClasses = 2
        
    for dataSetKey in myDataSets:
        [x_train,y_train,x_test,y_test] = ProcessData.obtainTrainingAndTestingData(dataSetKey)
        [x_trainScaled,x_testScaled] = ProcessData.scaleData(x_train,x_test) # Scale the data with StandardScaler() of scikit-learn

        print("Compiling ROCKET functions (once only)".ljust(80 - 5, "."), end = "", flush = True)
        
        _ = generate_kernels(20, 10)
        apply_kernels(np.zeros_like(x_trainScaled)[:, 1:], _)
        
        print("Done.")
    
        x_trainScaledWithoutValidationSet,x_validTrainScaled,y_trainWithoutValidationSet,y_validTrain = train_test_split(x_trainScaled,y_train,test_size=0.1,shuffle= True)
        
        # -- generate kernels ------------------------------------------------------
        kernels = generate_kernels(x_trainScaledWithoutValidationSet.shape[1], numberOfKernels)
        
        # -- train -----------------------------------------------------------------
        
        time_a = time.perf_counter()
        model, f_mean, f_std = train(x_trainScaledWithoutValidationSet,y_trainWithoutValidationSet,x_validTrainScaled,y_validTrain,kernels,numberOfKernels * 2,numberOfClasses)
        time_b = time.perf_counter()
    
        time_training_seconds = time_b - time_a
        print('Training took: ' + str(time_training_seconds) + ' s...')
        
        # -- test ------------------------------------------------------------------
    
        # normalise and transform test data
        X_test_transform = apply_kernels(x_testScaled, kernels)
        X_test_transform = (X_test_transform - f_mean) / f_std # normalise transformed features
        
        # predict
        model.eval()
        Y_test_predictions = model(torch.FloatTensor(X_test_transform))
        
        print(Y_test_predictions)
        Y_test_predictionsDetached = Y_test_predictions.detach().numpy()
        print(Y_test_predictionsDetached)
        
        del model
        K.clear_session()
        gc.collect()
```
The output of the last two print statements is something along those lines:

```
tensor([[ 8.5430e-01, -8.5430e-01],
        [ 3.8321e-01, -3.8321e-01],
        [ 1.1046e+00, -1.1046e+00],
        [ 9.2527e-02, -9.2526e-02],
...], grad_fn=<AddmmBackward>)
[[ 8.54303002e-01 -8.54302824e-01]
 [ 3.83211762e-01 -3.83211493e-01]
 [ 1.10462642e+00 -1.10462606e+00]
 [ 9.25266296e-02 -9.25263613e-02]
...]
```
I have three questions about this:

1.) Is the line 
`[x_trainScaled,x_testScaled] = ProcessData.scaleData(x_train,x_test) # Scale the data with StandardScaler() of scikit-learn`
 actually needed or should I use unscaled data as input for the methods?
2.) I'm facing a binary classification scenario and would like to compute the F1_scores from the predictions and the ground truth. For example my ground truth might look like [1,1,1,1] and the prediction could be [0,1,0,1]. How would I transform the values of my "Y_test_predictionsDetached" array to binary predictions?
3.) Are there any (other) obvious mistakes in my code?

Once again: Thank you very much for staying with me on this issue. It's very much appreciated. :)

**Comment by angus924** [comment]
Thank you for taking the time to try this.

In response to your questions:

> Is the line [...] actually needed [...]?

Short answer: no, you don't need this line.

Long answer: normalisation of the input time series is needed, unless your input time series are already normalised such that each time series has a mean of zero and a standard deviation of one (some datasets come already normalised).  However, if you are using the `train(...)` function from [`reproduce_experiments_scalability.py`](https://github.com/angus924/rocket/blob/master/code/reproduce_experiments_scalability.py), it is already doing the normalisation for you (for the training and validation data).  You will, however, have to perform normalisation on your test data.  StandardScaler from scikit-learn does the 'wrong kind' of normalisation / scaling in this case (per-feature normalisation).  Standard practice for time series is to normalise such that each time series has a mean of zero and a standard deviation of one.  In your case, I think something like this should work for your test data:

```python
x_test = (x_test - x_test.mean(axis = 1, keepdims = True)) / (x_test.std(axis = 1, keepdims = True) + 1e-8)
```

You will see that this is how we do input normalisation in [`reproduce_experiments_scalability.py`](https://github.com/angus924/rocket/blob/master/code/reproduce_experiments_scalability.py#L290).

> How would I transform the values [...] to binary predictions?

This is a more general PyTorch question; hopefully I can say something sensible here.  I believe that the values you are seeing for `Y_test_predictions` / `Y_test_predictionsDetached` are the raw output values produced by the model (pre-softmax).  Each row represents a (test) example, and each column represents a class (two columns for two classes).  If you want to convert these values into predictions, then you can do something like the following:

```python
binary_predictions = Y_test_predictions.max(1)[1]
```

Calling `.max(1)` produces the maximum value (for each row / test example)—actually, it produces two sets of values: the maximum value (per row) and the column index for the maximum value (i.e., argmax).  So, by doing `.max(1)[1]`, we are finding the class (column) with the biggest output value, i.e., the class predicted by the model (for all examples).  (If, instead, you want to get the probabilities produced by the model, you can do something like `s = nn.Softmax(dim = 1); probabilities = s(Y_test_predictions)`.)

You can compare the predictions with the test targets / labels as follows:

```python
test_accuracy = (Y_test_predictions.max(1)[1].numpy() == Y_test).mean()
```

You will see that we do something similar in [`reproduce_experiments_scalability.py`](https://github.com/angus924/rocket/blob/master/code/reproduce_experiments_scalability.py#L298).

> Are there any (other) obvious mistakes in my code?

I can't see any obvious problems, but it can be hard to catch subtle issues just by looking at the code.

I suggest trying it out and seeing if it produces sensible results for you.  If not, let me know, and we'll look into it and see what we can do.

Two things I did notice (not mistakes, just suggestions):

1. You could try adding the argument `cache_size = len(x_trainScaledWithoutValidationSet)` to your call to the `train(...)` function.  I don't know how many epochs are running (you could edit the `train(...)` function to add `print(f"epoch = {epoch}")` or something like that to find out), but caching the transform for the whole dataset will speed things up a lot for multiple epochs of training (the default setting I think is to only cache 2<sup>14</sup> or about 16K examples).
2. You can probably move the precompiling code outside your `for dataSetKey in myDataSets` loop, so that precompiling is done only once (actually, it won't be done more than once anyway, but there's no need for the code to be rerun for each dataset), or have a flag and an `if` statement, to check if precompiling has been done, and only precompile on the first run / dataset.  Also, your line `apply_kernels(np.zeros_like(x_trainScaled)[:, 1:], _)` for precompiling should be just `apply_kernels(np.zeros_like(x_trainScaled), _)`, as you've already split the data into X / features and Y / targets.  (Forcing Numba to precompile can be a bit fiddly, but it's not really important.)  You can tell if it is working correctly by timing the precompile code: the first time you run the code precompiling should take maybe 2 or 3 seconds, but after that the same code should execute almost instantaneously.  If precompiling is not working, it will take 2 or 3 seconds each time (as it is recompiling every time, which shouldn't happen).

I hope this helps, and please let me know if you have further issues or questions.

Best,

Angus.

**Comment by Huii** [comment]
Hi Angus,

Thank you very much for your answer. I hope you can one day add all this information to the documentation. It would help people new to ROCKET tremendously in the future.

For now, I implemented all your suggestions and I will let my machine run over the weekend to see how well the models perform with 10,000 kernels.

All the best,
Huii
